namer/namer.cc

```diff
@@ -15,392 +18,730 @@
+    u4 idx() const {
+        // Mask top bits.
+        return _id & 0x0FFF;
+    }
 
-        if (parsedArg.default_) {
-            argInfo.flags.isDefault = true;
-            localExpr = ast::MK::OptionalArg(parsedArg.loc, move(localExpr), move(parsedArg.default_));
-        }
+    FoundClassRef &klassRef(FoundNames &foundNames);
+    const FoundClassRef &klassRef(const FoundNames &foundNames) const;
 
-        if (parsedArg.keyword) {
-            argInfo.flags.isKeyword = true;
-        }
-        if (parsedArg.block) {
-            argInfo.flags.isBlock = true;
-        }
-        if (parsedArg.repeated) {
-            argInfo.flags.isRepeated = true;
-        }
+    FoundClass &klass(FoundNames &foundNames);
+    const FoundClass &klass(const FoundNames &foundNames) const;
 
-        return move(localExpr);
+    FoundMethod &method(FoundNames &foundNames);
+    const FoundMethod &method(const FoundNames &foundNames) const;
+
+    FoundFieldOrVariable &fieldOrVariable(FoundNames &foundNames);
+    const FoundFieldOrVariable &fieldOrVariable(const FoundNames &foundNames) const;
+
+    FoundTypeMember &typeMember(FoundNames &foundNames);
+    const FoundTypeMember &typeMember(const FoundNames &foundNames) const;
+
+    core::SymbolRef symbol(const FoundNames &foundNames) const;
+};
+
+struct FoundClassRef final {
+    core::NameRef name;
+    core::LocOffsets loc;
+    // If !owner.exists(), owner is determined by reference site.
+    FoundNameRef owner;
+};
+
+struct FoundClass final {
+    FoundNameRef owner;
+    FoundNameRef klass;
+    core::LocOffsets loc;
+    core::Loc declLoc;
+    ast::ClassDef::Kind classKind;
+};
+
+struct FoundFieldOrVariable final {
+    FoundNameRef owner;
+    FoundNameRef klass;
+    core::NameRef name;
+    core::LocOffsets asgnLoc;
+    core::LocOffsets lhsLoc;
+    bool isTypeAlias = false;
+};
+
+struct FoundTypeMember final {
+    FoundNameRef owner;
+    core::NameRef name;
+    core::LocOffsets asgnLoc;
+    core::LocOffsets nameLoc;
+    core::NameRef varianceName;
+    bool isFixed = false;
+    bool isTypeTemplete = false;
+    bool tooManyArgs = false;
+};
+
+struct FoundMethod final {
+    FoundNameRef owner;
+    core::NameRef name;
+    core::LocOffsets loc;
+    core::Loc declLoc;
+    ast::MethodDef::Flags flags;
+    vector<ast::ParsedArg> parsedArgs;
+};
+
+struct Modifier {
+    NameKind kind;
+    FoundNameRef owner;
+    core::LocOffsets loc;
+    // The name of the modification.
+    core::NameRef name;
+    // For methods: The name of the method being modified.
+    core::NameRef methodName;
+};
+
+class FoundNames final {
+    vector<FoundNameRef> _definitions;
+    vector<FoundClassRef> _klassRefs;
+    vector<FoundClass> _klasses;
+    vector<FoundMethod> _methods;
+    vector<FoundFieldOrVariable> _fieldOrVariables;
+    vector<FoundTypeMember> _typeMembers;
+    vector<core::SymbolRef> _symbols;
+    vector<Modifier> _modifiers;
```

Comment:
I originally had one vector of definitions and one with modifiers and used a `union` to differentiate different types of names, but it became complicated because I couldn't use default move constructors since method contains a `vector` which has a user-defined move constructor.

I then switched to this one-vector-per-definition-type design, and introduced a `FoundNameRef` class that encodes the type of ref and its vector in a single u4. 

This design will be really easy to cache, but there are a lot of vectors! I'm not sure if there's a better design I could use.

Reply:
Having more separate vectors seems easier to me(unless they crosstalk)